### PR TITLE
 Implement new multigrid transfer with MatrixFree infrastructure

### LINKED ISSUE #14968: Improve performance of setup of `MGTwoLevelTransfer`/`MGTransferGlobalCoarsening`
Already in two projects, I observed that `MGTwoLevelTransfer::reinit()` is takes a long time for polynomial transfer. More precisely, I observed that the case https://github.com/dealii/dealii/blob/a1b1356b0315c70e237050e779216ebdc26affcd/include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h#L3240-L3248 with uniform polynomial degree takes much longer to set up than the geometric transfer, which is, given the simplicity in case of the same mesh, surprising. Obviously, the code is written for the general case, but there might be room for some improvements. Maybe we should also add a benchmark similar to https://github.com/dealii/dealii/blob/master/tests/performance/timing_step_37.cc but using some combination of adaptive meshes, global coarsening and p-multigrid to ensure that these steps get done reasonably well.

### PR DESCRIPTION
**Pull request by kronbichler**
This implements a variant of `MGTwoLevelTransfer` that uses an already built `MatrixFree` object to define the interpolation and restriction operations, rather than setting up indices again. This will fix the main issue by #14968. On my system, I see a performance improvement in the setup (clearly), but also during the multigrid cycle because we transfer less memory: Running a loop with the matrix-free ingredients for the transfer will use the same DoF indices as the matrix-free operator evaluation. For cases where these indices fit partly into cache memory, this will reduce the memory access.

Of course, the proposed approach only works for p-multigrid, as h-multigrid will use different triangulations (or levels) and thus we cannot establish a connection between the matrix-free objects.

This PR is a first step: I would like to also implement the multi-component case with this approach, but I first wanted to hear opinions. We could also extend it to the hp case, but that is a bit more work. ~~Besides the general approach, a question concerns the name. I chose `MGTwoLevelTransferPrebuilt` but this is maybe misleading because we have `MGTransferPrebuilt` that sets up matrices. We could use `MGTwoLevelTransferMatrixFree` to indicate that we use `MatrixFree` objects to define the transfer ingredients. It has the possibility for confusion because also the other concept `MGTwoLevelTransfer` uses matrix-free ingredients (but rebuilding the DoF indices instead of using them from `MatrixFree`), so I initially opted against it. Are there other suggestions?~~

Fixes #14968.

### CODE REVIEW

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 1]
```diff
@@ -3831,6 +3830,491 @@ MGTwoLevelTransfer<dim, VectorType>::memory_consumption() const
...
+        for (unsigned int i = 0; i < fe_fine_1d->dofs_per_line; ++i)
+          renumbering_fine[i + fe_fine_1d->n_dofs_per_vertex()] =
+            GeometryInfo<1>::vertices_per_cell *
+              fe_fine_1d->n_dofs_per_vertex() +
+            i;
+        if (fe_fine_1d->n_dofs_per_vertex() > 0)
+          renumbering_fine[fe_fine_1d->n_dofs_per_cell() -
+                           fe_fine_1d->n_dofs_per_vertex()] =
+            fe_fine_1d->n_dofs_per_vertex();
+      }
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:4039]: All of this part (and some 90 lines below) repeat more or less what we did for the other transfer. Also, there is some repetition for the weights. If this PR finds approval, I will unify the code (either here or in a separate PR), but I wanted to get the general concept approved before I start refactoring.

**Comment by tjhei** [comment]
Thank you for working on this. I haven't looked at the code yet. Would it be feasible to put it into the same class (with an overloaded constructor)?
I find the "prebuilt" name confusing for the reasons you describe.

**Comment by kronbichler** [comment]
Yes, we could definitely move everything into the same class, and choose different implementations internally. This resolves the biggest question I had regarding the names, and I think we internally need to make sure that whatever we do is structured nicely. I will give this a try early next week.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h (hunk 1 above)
Comment (kronbichler) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h]: I now pushed a new version that unifies the two places, making the added code much nicer.

**Comment by kronbichler** [comment]
I now moved to the code suggested by @tjhei. I like this much better, because now we can have a single object, just initialize it differently. The `prolongate_and_add_internal`, `restrict_and_add_internal` functions are now a bit uglier because we need to implement two different evaluation concepts that are switched via an `if`; we might want to have a look at that later if we find it too confusing.

**Comment by kronbichler** [comment]
This PR is ready for review, I finally got all tests passing. Sorry that it took me a while, there were some subtle issues that I faced when separating out the code into modular functions that I missed initially. The PR looks pretty massive with around 1k lines of code changed in the `mg_transfer_global_coarsening.h/templates.h` files, but quite a bit of code has just moved around or is indented differently. Looking at the change without whitespace helps.

**Review by tjhei — APPROVED** [review]
Very nice. I think the interface is quite easy to use this way.

Are there any examples/ that we need to update?

**Comment by tjhei** [comment]
Oh, And this should go into 9.6.

**Comment by tjhei** [comment]
OK. Good to go from my side. Let's see if anybody else wants to take a look.

**Comment by kronbichler** [comment]
There were no new comments - let us merge this now to give it some time with the testers. (I will see if I can make the extension to multiple components easily in the next day.)

**Pull request merged**